src/index.d.ts

```diff
@@ -0,0 +1,314 @@
+type PromiseExecutor<T> = (
+  resolve: (value?: T | PromiseLike<T>) => void,
+  reject: (reason?: any) => void
+) => void;
+
+interface PromiseFulfilledResult<T> {
+  status: 'fulfilled';
+  value: T;
+}
+
+interface PromiseRejectedResult {
+  status: 'rejected';
+  reason: any;
+}
+
+type PromiseSettledResult<T> =
+  | PromiseFulfilledResult<T>
+  | PromiseRejectedResult;
+
+interface CancelablePromiseType<T> {
+  /**
+   * Attaches callbacks for the resolution and/or rejection of the Promise.
+   * @param onfulfilled The callback to execute when the Promise is resolved.
+   * @param onrejected The callback to execute when the Promise is rejected.
+   * @returns A Promise for the completion of which ever callback is executed.
+   */
+  then<TResult1 = T, TResult2 = never>(
+    onfulfilled?:
+      | ((value: T) => TResult1 | PromiseLike<TResult1>)
+      | undefined
+      | null,
+    onrejected?:
+      | ((reason: any) => TResult2 | PromiseLike<TResult2>)
+      | undefined
+      | null
+  ): CancelablePromiseType<TResult1 | TResult2>;
+
+  /**
+   * Attaches a callback for only the rejection of the Promise.
+   * @param onrejected The callback to execute when the Promise is rejected.
+   * @returns A Promise for the completion of the callback.
+   */
+  catch<TResult = never>(
+    onrejected?:
+      | ((reason: any) => TResult | PromiseLike<TResult>)
+      | undefined
+      | null
+  ): CancelablePromiseType<T | TResult>;
+
+  /**
+   * Attaches a callback that is invoked when the Promise is settled (fulfilled or rejected). The
+   * resolved value cannot be modified from the callback.
+   * @param onfinally The callback to execute when the Promise is settled (fulfilled or rejected).
+   * @returns A Promise for the completion of the callback.
+   */
+  finally(
+    onfinally?: (() => void) | undefined | null
+  ): CancelablePromiseType<T>;
+
+  cancel(onCancel: () => void): void;
+}
+
+interface CancelablePromiseConstructor {
```

Comment:
I copy/pasted `Promise` types from `typescript/lib` and changed returned types, I don't know if there is a better way to type this 🤷‍♀️

Reply:
I guess you can 
```
interface CancelablePromise extends Promise {
  _promise: Promise<any>;
  cancel()...
}
```

Reply:
Why you don't type directly in index.ts ?

Reply:
I see your issue since you return CancelablePromise instead of Promise in every methods. For now I don't see a cleaner solution since there is nothing common between Promise and CancelablePromise in the end (race, all, ... returns Cancelable, same for then, catch, ...)

Reply:
indeed I can migrate sources to typescript, I think I will

Reply:
finally I won't... it's so painful to write the class in TS, I don't know exactly how to handle multiple signatures such as `all<T1>()`, `all<T1, T2>()`, etc

Reply:
Yeah, i don't see a better way to do it. Since it's typed, it's ok. Maybe someone will come to this repository and bring us light on this :)